decoder.go

```diff
@@ -0,0 +1,133 @@
+package drum
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"log"
+	"os"
+	"strings"
+)
+
+// DecodeFile decodes the drum machine file found at the provided path
+// and returns a pointer to a parsed pattern which is the entry point to the
+// rest of the data.
+func DecodeFile(path string) (*Pattern, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Pattern{}
+	if err := NewDecoder(file).Decode(p); err != nil {
+		log.Fatal(err)
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Decoder reads and decodes splice patterns from an input stream
+type Decoder struct {
+	io.Reader
+}
+
+// NewDecoder returns a new decoder that reads from r
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{r}
+}
+
+// Decode reads the input stream and decodes it's values into a Pattern
+func (d *Decoder) Decode(p *Pattern) error {
+	n, err := d.spliceHeaderInfo()
+	if err != nil {
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			return errors.New("unable to decode .splice file header: " + err.Error())
+		}
+	}
+
+	if err := d.decodeBody(p, n); err != nil {
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			return errors.New("unable to decode .splice format: " + err.Error())
+		}
+	}
+	return nil
+}
+
+// spliceHeaderInfo will read the inital bites
+// and returns the size of the file contents or error if it's
+// unable check the SPLICE portion or read the headers
+func (d *Decoder) spliceHeaderInfo() (int64, error) {
+	hdr := make([]byte, 6)
+	if _, err := io.ReadFull(d, hdr); err != nil {
+		return 0, err
+	}
+	if string(hdr) != "SPLICE" {
+		return 0, errors.New("unable to decode non SPLICE files")
+	}
+
+	var size int64
+	if err := binary.Read(d, binary.BigEndian, &size); err != nil {
```

Comment:
The choice of an int64 here is odd considering we're parsing untrusted input. What if the size you read is negative?